src/internal/drawImage.js

```diff
@@ -6,9 +18,169 @@
  * @returns {void}
  * @memberof Internal
  */
-export default function (enabledElement, invalidated = false) {
-  enabledElement.needsRedraw = true;
+
+
+export function drawImage (enabledElement, invalidated = false) {
+  registerVisibilityChange();
+
+  if (!enabledElement || !enabledElement.canvas || (!enabledElement.image && enabledElement.layers.length < 1)) {
+    return;
+  }
+  
+  const element = enabledElement.element;
+  
   if (invalidated) {
     enabledElement.invalid = true;
   }
+
+  if (!elementsDrawMap.has(element)) {
+    elementsDrawMap.set(element, { 
+      needsRedraw: false, 
+      callbacks: new Set()
+    });
+  }
+
+  elementsDrawMap.get(element).needsRedraw = true;
+
+  rafStopTime = performance.now() + rafStopMaxTime;
+  if (!rafId) {
+    rafId = requestAnimationFrame(step);
+  }
+}
+
+function step () {
+  elementsDrawMap.forEach((entry, element) => {
+    if (!element || !entry || !entry.needsRedraw) {
+      return;
+    }
+    
+    const hasCallbacks = entry.callbacks.size > 0;
+    
+    entry.needsRedraw = hasCallbacks;
+
+    let enabledElement;
+
+    try { 
+      enabledElement = getEnabledElement(element); 
+    } catch (err) { 
+      return; 
+    }
+    
+    const eventDetails = {
+      enabledElement,
+      timestamp: Date.now()
```

Comment:
Why not use the RAF timestamp? I don't have strong feeling about this; I am just wondering.
This means that time-based CINE synchronization wont be in lock step? One viewport could be ahead of another if it's draw function is called later in the same frame?

Reply:
I have never used the PRE_RENDER event and never seen any code that does so I am not sure how best to use it. If you feel sending the RAF timestamp is better I'm happy to change it ?

The better implementation for time-based CINE synchronization like tools to call addDrawCallback  and then on the callback calculate what image needs to be displayed based on the time that passed since the beginning of playing (which they keep account of internally) this is much more accurate and does not rely on the RAF timing at all. 

Something like:
```
function rafCallback(enabledElement) {
	const delta = (performance.now() - framesClipData.playStartTime) % (framesClipData.loopPlayTime / framesClipData.speed);
	const newFrame = Math.floor(delta / (framesClipData.msPerFrame / framesClipData.speed));
	if (newFrame===framesClipData.currFrame) return;

	framesClipData.currFrame = newFrame;
  if (framesClipData.currFrame > framesClipData.numberOfFrames) framesClipData.currFrame-=framesClipData.numberOfFrames;
	const newImage = framesClipData.images[framesClipData.currFrame];
	if (newImage) enabledElement.image=newImage;
}

function play (element) {
  cornerstone.addDrawCallback(element, rafCallback);
}

function stop (element) {
	cornerstone.removeDrawCallback(element, rafCallback);
}
```

Note that rafCallback doesn't need to call any draw image function (though there's no issue if it does), all it needs to do is change enabledElement.image. There is no need for any setTimeout/setInterval etc ..

Reply:
I might be the only one that ever used `PRE_RENDER`. I no longer have access to that codebase, but I am not sure they still track upstream version of cornerstone. It might be possible to remove it to reclaim some performance; it's just a bit risky..

I don't have strong opinion on the timing. The tradeoff is between 1) always draw the frame with the most up to date time (some viewport could be at time t and other at t+1 in the same browser frame because they are called later) 2) ensure that all viewport are in sync at time t for a given browser frame.